docs/architecture/adr-070-unordered-account.md

```diff
@@ -2,214 +2,134 @@
-However, we'll note a few important layers of protection and mitigation:
-
-* Assuming CometBFT is used as the underlying consensus engine and a non-noop mempool
-  is used, CometBFT will reject the duplicate for you.
-* For applications that leverage ABCI++, `ProcessProposal` should evaluate and reject
-  malicious proposals with duplicate transactions.
-* For applications that leverage their own application mempool, their mempool should
-  reject the duplicate for you.
-* Finally, worst case if the duplicate transaction is somehow selected for a block
-  proposal, 2nd and all further attempts to evaluate it, will fail during `DeliverTx`,
-  so worst case you just end up filling up block space with a duplicate transaction.
+We facilitate replay protection by storing the unordered sequence, in the Cosmos SDK KV store. Upon transaction ingress, we check if the transaction's unordered
+sequence exists in state, or if the TTL value is stale, i.e. before the current block time. If so, we reject it. Otherwise,
+we add the unordered sequence to state.
 
-```golang
-type TxHash [32]byte
+The state is evaluated during `PreBlocker`. All transactions with an unordered sequence earlier than the current block time
+will be deleted.
 
-const PurgeLoopSleepMS = 500
+```golang
+package unorderedtx
 
-// UnorderedTxManager contains the tx hash dictionary for duplicates checking,
-// and expire them when block production progresses.
-type UnorderedTxManager struct {
-  // blockCh defines a channel to receive newly committed block time
-  blockCh chan time.Time
+import (
+  "fmt"
+  "time"
 
-  mu sync.RWMutex
-	// txHashes defines a map from tx hash -> TTL value, which is used for duplicate
-	// checking and replay protection, as well as purging the map when the TTL is
-	// expired.
-	txHashes map[TxHash]time.Time
-}
+  sdk "github.com/cosmos/cosmos-sdk/types"
 
-func NewUnorderedTxManager() *UnorderedTxManager {
-  m := &UnorderedTxManager{
-		blockCh:  make(chan time.Time, 16),
-		txHashes: make(map[TxHash]time.Time),
-  }
+  "cosmossdk.io/collections"
+  "cosmossdk.io/core/store"
+)
 
- return m
-}
+const (
+  // DefaultMaxTimeoutDuration defines the default maximum duration an unordered transaction
+  // can set.
+  DefaultMaxTimeoutDuration = time.Minute * 10
+)
 
-func (m *UnorderedTxManager) Start() {
-  go m.purgeLoop()
-}
+type UnorderedSequence string
 
-func (m *UnorderedTxManager) Close() error {
-  close(m.blockCh)
-  m.blockCh = nil
-  return nil
+func NewUnorderedSequence(addr string, timestamp uint64) UnorderedSequence {
+  return UnorderedSequence(fmt.Sprintf("%d/%s", timestamp, addr))
 }
 
-func (m *UnorderedTxManager) Contains(hash TxHash)  bool{
-  m.mu.RLock()
-  defer m.mu.RUnlock()
-
-  _, ok := m.txHashes[hash]
-  return ok
+// Manager contains the tx hash dictionary for duplicates checking, and expire
+// them when block production progresses.
+type Manager struct {
+  kvStore store.KVStoreService
 }
```

Comment:
What module do we want to have this in? All KV store keys need a module. x/auth seems the most logical, but we could also have a small dedicated x/auth/unordered module.

Also, I think we can just use `collections.Set` to deal with state here.

Reply:
> What module do we want to have this in?

are there tradeoffs between doing one or the other?

> Also, I think we can just use collections.Set to deal with state here.

i was toying around with this, but was having trouble figuring out how to iterate from nil to a partial value of `<block_time/`.

Reply:
i have a PoC where i've made unordered a thing managed directly by x/auth store/keeper. you can check it out here: https://github.com/cosmos/cosmos-sdk/pull/24010

i also switched to using collections after playing with the API more.

Reply:
makes sense - i did find it cumbersome going back and forth between this and the draft. throwing some simple python in here would be easier to deal with